connexion/middleware/routing.py

```diff
@@ -0,0 +1,172 @@
+from connexion.operations import AbstractOperation, make_operation
+from connexion.resolver import Resolver
+
+CONNEXION_CONTEXT = 'CONNEXION_CONTEXT'
+
+
+_scope_receive_send: ContextVar[tuple] = ContextVar('SCOPE_RECEIVE_SEND')
+
+
+class MiddlewareResolver(Resolver):
+
+    def __init__(self, call_next: t.Callable) -> None:
+        """Resolver that resolves each operation to the provided call_next function."""
+        super().__init__()
+        self.call_next = call_next
+
+    def resolve_function_from_operation_id(self, operation_id: str) -> t.Callable:
+        return self.call_next
+
+
+class RoutingMiddleware(AppMiddleware):
+
+    def __init__(self, app: ASGIApp) -> None:
+        """Middleware that resolves the Operation for an incoming request and attaches it to the
+        scope.
+
+        :param app: app to wrap in middleware.
+        """
+        self.app = app
+        # Pass unknown routes to next app
+        self.router = Router(default=self.default_fn)
+
+    def add_api(
+            self,
+            specification: t.Union[pathlib.Path, str, dict],
+            base_path: t.Optional[str] = None,
+            arguments: t.Optional[dict] = None,
+            **kwargs
+    ) -> None:
+        """Add an API to the router based on a OpenAPI spec.
+
+        :param specification: OpenAPI spec as dict or path to file.
+        :param base_path: Base path where to add this API.
+        :param arguments: Jinja arguments to replace in the spec.
+        """
+        kwargs.pop("resolver", None)
+        resolver = MiddlewareResolver(self.create_call_next())
+        api = MiddlewareAPI(specification, base_path=base_path, arguments=arguments,
+                            resolver=resolver, default=self.default_fn, **kwargs)
+        self.router.mount(api.base_path, app=api.router)
+
+    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
+        """Route request to matching operation, and attach it to the scope before calling the
+        next app."""
+        if scope["type"] != "http":
+            await self.app(scope, receive, send)
+            return
+
+        _scope_receive_send.set((scope.copy(), receive, send))
+
+        # Needs to be set so starlette router throws exceptions instead of returning error responses
+        scope['app'] = self
+        try:
+            await self.router(scope, receive, send)
+        except ValueError:
+            raise NotFoundProblem
+
+    async def default_fn(self, scope: Scope, receive: Receive, send: Send) -> None:
+        """Callback to call next app as default when no matching route is found."""
+        original_scope, *_ = _scope_receive_send.get()
+
+        api_base_path = scope.get('root_path', '')[len(original_scope.get('root_path', '')):]
+
+        if CONNEXION_CONTEXT not in original_scope:
+            original_scope[CONNEXION_CONTEXT] = {}
+
+        original_scope[CONNEXION_CONTEXT].update({
```

Comment:
FYI, the `dict.setdefault()` simplifies this a bit (although perhaps a bit less untuitive as it's less known):
```
original_scope.setdefault(CONNEXION_CONTEXT, {}).update({'api_base_path': api_base_path})
```

Reply:
Not a fan of the one-liner, but can indeed use `setdefault` for the first part.